File: 112.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    bool hasPathSum(TreeNode* root, int sum) {
        if(!root){
            return false;
        }
        sum -= root->val;
        if(sum == 0 && !root->left && !root->right){
            return true;
        }
        return hasPathSum(root->left,sum) || hasPathSum(root->right, sum);
    }
};

File: go-solutions/112.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func find(root *TreeNode, tmp int, sum *int) bool {
	if root == nil {
		return false
	}
	tmp += root.Val
	if (root.Left == nil) && (root.Right == nil) {
		if tmp == *sum {
			return true
		} else {
			return false
		}
	}
	return find(root.Left, tmp, sum) || find(root.Right, tmp, sum)
}

func hasPathSum(root *TreeNode, sum int) bool {
	return find(root, 0, &sum)
}

